src/Decompiler/Analysis/IndirectCallRewriter.cs

```diff
@@ -0,0 +1,213 @@
+﻿#region License
+/* 
+ * Copyright (C) 1999-2016 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Reko.Core;
+using Reko.Core.Code;
+using Reko.Core.Expressions;
+using Reko.Core.Types;
+using Reko.Typing;
+
+namespace Reko.Analysis
+{
+    /// <summary>
+    /// Try to rewrite indirect call statements to Applications using
+    /// user-defined data (e.g. global variables, parameters of procedures).
+    /// </summary>
+    class IndirectCallRewriter : ExpressionTypeAscender
+    {
```

Comment:
We want a unit test fixture for this new class. We need to validate that indirect calls where there is no type information available at all don't cause crashes.